### PR TITLE
Add Timeout to R2R

### PR DESCRIPTION
**Pull request by thestr4ng3r**

### CODE REVIEW

File: binr/r2r/run.c [hunk 1]
```diff
@@ -274,8 +281,16 @@ R_API R2RSubprocess *r2r_subprocess_start(
 	return NULL;
 }
 
-R_API void r2r_subprocess_wait(R2RSubprocess *proc) {
-	int r;
+R_API bool r2r_subprocess_wait(R2RSubprocess *proc, ut64 timeout_ms) {
+	struct timespec timeout_abs;
+	if (timeout_ms != UT64_MAX) {
+		clock_gettime(CLOCK_MONOTONIC, &timeout_abs);
```
Comment (radare) [inline binr/r2r/run.c]: missing space

File: binr/r2r/run.c [hunk 2]
```diff
@@ -274,8 +281,16 @@ R_API R2RSubprocess *r2r_subprocess_start(
...
+		timeout_abs.tv_nsec += timeout_ms * 1000000;
```
Comment (radare) [inline binr/r2r/run.c]: maybe use a constant define instead of 1000000

File: binr/r2r/run.c [hunk 3]
```diff
@@ -302,15 +317,32 @@ R_API void r2r_subprocess_wait(R2RSubprocess *proc) {
...
+		struct timeval *timeout = NULL;
+		if(timeout_ms != UT64_MAX)
+		{
+			struct timespec now;
+			clock_gettime(CLOCK_MONOTONIC, &now);
+			st64 usec_diff = ((st64)timeout_abs.tv_sec - now.tv_sec) * 1000000 + ((st64)timeout_abs.tv_nsec - now.tv_nsec) / 1000;
+			if (usec_diff <= 0) {
+				break;
+			}
+			timeout_s.tv_sec = usec_diff / 1000000;
```
Comment (radare) [inline binr/r2r/run.c]: same here

File: binr/r2r/run.c [hunk 4]
```diff
@@ -302,15 +317,32 @@ R_API void r2r_subprocess_wait(R2RSubprocess *proc) {
 			}
 		}
 		nfds++;
-		r = select (nfds, &rfds, NULL, NULL, NULL);
+
+		struct timeval timeout_s;
+		struct timeval *timeout = NULL;
+		if(timeout_ms != UT64_MAX)
```
Comment (radare) [inline binr/r2r/run.c]: missing sace and open brace {

File: binr/r2r/run.c [hunk 5]
```diff
@@ -334,12 +367,21 @@ R_API void r2r_subprocess_wait(R2RSubprocess *proc) {
...
+		}
 	}
 	if (r < 0) {
 		perror ("select");
 	}
+	return child_dead;
+}
+
+R_API void r2r_subprocess_kill(R2RSubprocess *proc) {
+	kill (proc->pid, SIGKILL);
```
Comment (radare) [inline binr/r2r/run.c]: #if __UNIX__

Comment (radare) [reply binr/r2r/run.c]: or use r_sys_kill

File: binr/r2r/run.c (hunk 2 above)
Comment (ret2libc) [reply binr/r2r/run.c]: yes please

Comment (ret2libc) [reply binr/r2r/run.c]: even better, some define like `MS_TO_NSEC` and `NSEC_TO_SEC`

**Comment by thestr4ng3r** [comment]
I was way too tired when I wrote this yesterday, which is also why I made it a draft.

**Comment by XVilka** [comment]
I suggest then the new status for such tests - not `[XX]` but `[TM]`, so it will be pretty clear from the logs.

**Comment by radare** [comment]
this is very important to support the fuzz testsuite, also add a commandline flag to specify the timeout value (maybe -t)

**Comment by radare** [comment]
what is TM? trademark?

**Comment by radare** [comment]
i prefer XX+ TM makes no sense

**Comment by XVilka** [comment]
You have to make a distinction between failed and timed out tests

**Comment by thestr4ng3r** [comment]
It should be enough to print it as timeout when showing the diff and then count it as XX.

**Comment by radare** [comment]
Agree with florian

**Comment by radare** [comment]
But i agree about the usefulness of having TM for timeouted tests

File: binr/r2r/run.c (hunk 5 above)
Comment (thestr4ng3r) [reply binr/r2r/run.c:395]: This is already in the `#else` branch of `#if __WINDOWS__` where tons of unix api is used so should be fine.

**Comment by thestr4ng3r** [comment]
This is what they look like now:
Screenshot from 2020-04-02 11-22-15

**Comment by radare** [comment]
i think 60s timeout is not enough for some fuzz tests but we can tweak it later when the fuzz tests run from here

**Comment by thestr4ng3r** [comment]
It's apparently not even enough for some regular tests.

**Comment by radare** [comment]
conflicts

File: binr/r2r/run.c [hunk 6]
```diff
@@ -2,6 +2,11 @@
 
 #include "r2r.h"
 
+#define NSEC_PER_SEC  1000000000
```
Comment (radare) [inline binr/r2r/run.c:5]: so many things that shouldnt be in r2r but in util :______

**Pull request merged**